# Processing and Analysis of 10X Multiomic data

## Overview

To familiarize myself with RNA and ATAC-seq analysis workflows, I opted to work raw data publicly available via the Allen Institute's <a href="https://brain-map.org/bkp">Brain Knowledge Platform</a>. These analyses were to be carried out on a Lenovo P51 Thinkpad with 8 cores and 64GB of memory, so to address computational resource limitations, I analyzed 10X Multiomic data generated from brainstem tissue in the common marmoset *Callithrix jacchus*.

Much of the following workflow and code are adapted from the following tutorials/vignettes: \
https://stuartlab.org/signac/articles/overview \
https://satijalab.org/seurat/articles/get_started_v5_new \
https://ngs101.com/tutorials/ \
https://github.com/mousepixels/sanbomics_scripts

## Pre-processing


## Cell Annotation

I annotated cells using the MapMyCells platform with the 10x Whole Mouse Brain Taxonomy (CCN20230722) as the reference.


## Differential Expression analysis


## Differential Accessibility analysis

## Sources

Dataset: Fenna M. Krienen, Kirsten M. Levandowski, Heather Zaniewski, Ricardo C.H. del Rosario, Margaret E. Schroeder, Melissa Goldman, Alyssa Lutservitz, Qiangge Zhang, Katelyn X. Li, Victoria F. Beja-Glasser, Jitendra Sharma, Tay Won Shin, Abigail Mauermann, Alec Wysoker, James Nemesh, Seva Kashin, Josselyn Vergara, Gabriele Chelini, Jordane Dimidschstein, Sabina Berretta, Ed Boyden, Steven A. McCarroll, Guoping Feng (2023) A Marmoset Brain Cell Census Reveals Persistent Influence of Developmental Origin on Neurons. [Dataset]. Available from https://assets.nemoarchive.org/dat-1je0mn3
